cmd/juju/controller/register.go

```diff
@@ -143,67 +255,70 @@ func (c *registerCommand) Run(ctx *cmd.Context) error {
 	payloadBytes, ok := secretbox.Open(nil, resp.PayloadCiphertext, &respNonce, &registrationParams.key)
 	if !ok {
-		return errors.NotValidf("response payload")
+		return errRet(errors.NotValidf("response payload"))
 	}
 	var responsePayload params.SecretKeyLoginResponsePayload
 	if err := json.Unmarshal(payloadBytes, &responsePayload); err != nil {
-		return errors.Annotate(err, "unmarshalling response payload")
-	}
-
-	// Store the controller and account details.
-	controllerDetails := jujuclient.ControllerDetails{
-		APIEndpoints:   registrationParams.controllerAddrs,
-		ControllerUUID: responsePayload.ControllerUUID,
-		CACert:         responsePayload.CACert,
-	}
-	if err := store.AddController(registrationParams.controllerName, controllerDetails); err != nil {
-		return errors.Trace(err)
-	}
-	accountDetails := jujuclient.AccountDetails{
-		User:            registrationParams.userTag.Canonical(),
-		LastKnownAccess: string(permission.LoginAccess),
-	}
-	if err := store.UpdateAccount(registrationParams.controllerName, accountDetails); err != nil {
-		return errors.Trace(err)
-	}
+		return errRet(errors.Annotate(err, "unmarshalling response payload"))
+	}
+	user := registrationParams.userTag.Canonical()
+	ctx.Infof("Initial password successfully set for %s.", friendlyUserName(user))
+	return jujuclient.ControllerDetails{
+			APIEndpoints:   registrationParams.controllerAddrs,
+			ControllerUUID: responsePayload.ControllerUUID,
+			CACert:         responsePayload.CACert,
+		}, jujuclient.AccountDetails{
+			User:            user,
+			LastKnownAccess: string(permission.LoginAccess),
+		}, nil
+}
 
-	// Log into the controller to verify the credentials, and
-	// list the models available.
-	models, err := c.listModelsFunc(store, registrationParams.controllerName, accountDetails.User)
+// updateController prompts for a controller name and updates the
+// controller and account details in the given client store.
+// It returns the name of the updated controller.
+func (c *registerCommand) updateController(
+	ctx *cmd.Context,
+	store jujuclient.ClientStore,
+	defaultControllerName string,
+	controllerDetails jujuclient.ControllerDetails,
+	accountDetails jujuclient.AccountDetails,
+) (string, error) {
+	// Check that the same controller isn't already stored, so that we
+	// can avoid needlessly asking for a controller name in that case.
+	all, err := store.AllControllers()
 	if err != nil {
-		return errors.Trace(err)
+		return "", errors.Trace(err)
 	}
-	for _, model := range models {
-		owner := names.NewUserTag(model.Owner)
-		if err := store.UpdateModel(
-			registrationParams.controllerName,
-			jujuclient.JoinOwnerModelName(owner, model.Name),
-			jujuclient.ModelDetails{model.UUID},
-		); err != nil {
-			return errors.Annotate(err, "storing model details")
+	for name, ctl := range all {
+		if ctl.ControllerUUID == controllerDetails.ControllerUUID {
+			// TODO(rogpeppe) Succeed but override the account details in this case?
```

Comment:
I think that would be sensible, if a little unusual. Can you please file a bug for that and reference?